src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java

```diff
@@ -129,20 +128,103 @@ protected void customize(Request exchange) {
 
     // Now set up the Bayeux client itself
     BayeuxClient client = new BayeuxClient(oAuthInfo.getInstanceURL() + DEFAULT_PUSH_ENDPOINT, transport);
-    client.handshake();
-
     return client;
   }
 
-  private void waitForHandshake(BayeuxClient client,
-                                long timeoutInMilliseconds, long intervalInMilliseconds) {
+  public void createSalesforceListener() throws Exception {
+    bayeuxClient = getClient(credentials);
+    bayeuxClient.getChannel(Channel.META_HANDSHAKE).addListener
+      ((ClientSessionChannel.MessageListener) (channel, message) -> {
+
+        boolean success = message.isSuccessful();
+        if (!success) {
+          String error = (String) message.get("error");
+          if (error != null) {
+            LOG.error("Error during HANDSHAKE: {}", error);
+            throw new RuntimeException("Error in meta handshake , errorMessage:" + error);
+          }
+          if (message.get("exception") instanceof Exception) {
+            Exception exception = (Exception) message.get("exception");
+            if (exception != null) {
+              LOG.error("Exception during HANDSHAKE", exception);
+              throw new RuntimeException("Error in meta handshake ", exception);
+            }
+          }
+        }
+      });
+
+    bayeuxClient.getChannel(Channel.META_CONNECT).addListener(
+      (ClientSessionChannel.MessageListener) (channel, message) -> {
+
+        boolean success = message.isSuccessful();
+        if (!success) {
+          String error = (String) message.get("error");
+          Map<String, Object> advice = message.getAdvice();
+
+          if (error != null) {
+            LOG.error("Error during CONNECT: {}", error);
+            LOG.debug("Advice during CONNECT: {}", advice);
+          }
+          // Error Codes Reference in Salesforce Streaming :
+          // https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/streaming_error_codes.htm
+          if (advice.get("reconnect").equals("handshake")) {
+            LOG.debug("Reconnecting to Salesforce Push Topic");
+            try {
+              reconnectToTopic();
+            } catch (Exception e) {
+              throw new RuntimeException("Error in reconnecting to salesforce ", e);
+            }
+          } else {
+            throw new RuntimeException("Error in meta connect , errorMessage:" + error + " Advice:" + advice);
+          }
+        }
+      });
+
+    bayeuxClient.getChannel(Channel.META_SUBSCRIBE).addListener(
+      (ClientSessionChannel.MessageListener) (channel, message) -> {
+
+        boolean success = message.isSuccessful();
+        if (!success) {
+          String error = (String) message.get("error");
+          if (error != null) {
+            LOG.error("Error during SUBSCRIBE: " + error);
+            throw new RuntimeException("Error in meta subscribe , errorMessage:" + error);
+          }
```

Comment:
what happens if the success = false and message doesn't contain `error`?

Reply:
The pipeline will continue to run as usual.